### PR TITLE
FIX stdio URL bug with proxyMode

### LINKED ISSUE #2920: MCPServer status URL shows /sse instead of /mcp when using stdio transport with streamable-http proxyMode
## Bug Description

When deploying an MCPServer with `transport: stdio` and `proxyMode: streamable-http`, the status URL incorrectly shows `/sse` instead of `/mcp`.

## Steps to Reproduce

Deploy an MCPServer like this:

```yaml
apiVersion: toolhive.stacklok.dev/v1alpha1
kind: MCPServer
metadata:
  name: arxiv
  namespace: toolhive-system
spec:
  image: ghcr.io/stacklok/dockyard/uvx/arxiv-mcp-server:0.3.1
  transport: stdio
  proxyMode: streamable-http
  proxyPort: 8080
```

## Expected Behavior

Status URL should show `/mcp` endpoint (e.g., `http://arxiv.toolhive-system.svc.cluster.local:8080/mcp`)

## Actual Behavior

Status URL shows `/sse` endpoint (e.g., `http://arxiv.toolhive-system.svc.cluster.local:8080/sse#arxiv`)

## Root Cause

In `pkg/transport/url.go`, the `GenerateMCPServerURL` function treats `stdio` the same as `sse`:

```go
isSSE := transportType == types.TransportTypeSSE.String() || transportType == types.TransportTypeStdio.String()
```

The MCPServer controller at `cmd/thv-operator/controllers/mcpserver_controller.go:395-401` only passes the `Transport` field, ignoring `ProxyMode`:

```go
mcpServer.Status.URL = transport.GenerateMCPServerURL(
    mcpServer.Spec.Transport,   // <-- Uses Transport, NOT ProxyMode
    host,
    int(mcpServer.GetProxyPort()),
    mcpServer.Name,
    "",
)
```

## Impact

| Transport | ProxyMode | Actual Endpoint | Status Shows | Match? |
|-----------|-----------|-----------------|--------------|--------|
| `stdio` | `sse` | `/sse` | `/sse` | ✅ |
| `stdio` | `streamable-http` | `/mcp` | `/sse` | ❌ **BUG** |
| `sse` | (ignored) | `/sse` | `/sse` | ✅ |
| `streamable-http` | (ignored) | `/mcp` | `/mcp` | ✅ |

Since the default `proxyMode` is `streamable-http`, **this bug affects the default stdio configuration**.

## Proposed Fix

Add a `proxyMode` parameter to `GenerateMCPServerURL` and update all callers:

1. **`pkg/transport/url.go`** - Add `proxyMode` parameter:
   - For `stdio` transport: use `proxyMode` to determine endpoint
   - For `stdio` with empty `proxyMode`: default to `streamable-http` (matches CRD default)
   - For non-stdio transports: ignore `proxyMode`

2. **Update call sites** (5 locations):
   - `cmd/thv-operator/controllers/mcpserver_controller.go:395`
   - `pkg/runner/runner.go:283`
   - `pkg/workloads/manager.go:1392`
   - `pkg/workloads/types/types.go:75`
   - `pkg/vmcp/workloads/k8s.go:141`

3. **Update tests** in `pkg/transport/url_test.go`

## Labels

bug, operator

### PR DESCRIPTION
**Pull request by slyt3**
## Problem

So deploying an MCPServer with `transport: stdio` and `proxyMode: streamable-http`, the status URL shows incorectly returns `/sse` endpoint instead of `/mcp` 

### Example
```yaml
apiVersion: toolhive.stacklok.dev/v1alpha1
kind: MCPServer
metadata:
  name: arxiv
  namespace: toolhive-system
spec:
  image: ghcr.io/stacklok/dockyard/uvx/arxiv-mcp-server:0.3.1
  transport: stdio
  proxyMode: streamable-http
  proxyPort: 8080
```

**Expected status URL:**
```
http://arxiv.toolhive-system.svc.cluster.local:8080/mcp
```

**Actual status URL:**
```
http://arxiv.toolhive-system.svc.cluster.local:8080/sse#arxiv
```

---

## Root Cause

The bug is in `pkg/transport/url.go` 

This code is old and it shows that its treating stdio same as sse and ignoring proxyMode
```go
isSSE := transportType == types.TransportTypeSSE.String() || 
         transportType == types.TransportTypeStdio.String()
```

So thinking logically we should assume that stdio always use SSE endpoint but stdio can use it either
- `proxyMode: sse` → `/sse` endpoint
- `proxyMode: streamable-http` → `/mcp` endpoint (default)

And the controller at `cmd/thv-operator/controllers/mcpserver_controller.go:395` only passes `Transport`, never `ProxyMode`:

OLD CODE TOO missing proxyMode parameter
```go
mcpServer.Status.URL = transport.GenerateMCPServerURL(
    mcpServer.Spec.Transport,
    host,
    int(mcpServer.GetProxyPort()),
    mcpServer.Name,
    "",
)
```

---

## Solution

So I Added `proxyMode` parameter to `GenerateMCPServerURL()` function:
```go
func GenerateMCPServerURL(transportType string, proxyMode string, host string, 
    port int, containerName, remoteURL string) string
```

### Logic:

1. **For stdio transport:** Use `proxyMode` to  easily determine endpoint
   - `proxyMode: "streamable-http"` → `/mcp`
   - `proxyMode: "sse"` → `/sse#name`
   - `proxyMode: ""` (empty) → defaults to `/mcp` (matches CRD default)

2. **For other transports:** JUST Ignore `proxyMode` its behaviour doesnt change xD
   - `sse` transport → always `/sse#name`
   - `streamable-http` transport → always `/mcp`

### Type Conversions

Added `string()` conversions where `ProxyMode` is a typed string (was stuck on this one for whole night, was going crazy i just needed to convert ProxyMode string to function call, so proxyMode is defined as  type ProxyMode string:
```go
string(r.Config.ProxyMode)  
```

---

### Unit Tests (15/15 passing)
```bash
$ go test ./pkg/transport/... -v -run TestGenerateMCPServerURL
```

Key test cases covering the bug:

| Test | Transport | ProxyMode | Expected | Result |
|------|-----------|-----------|----------|--------|
| **Bug case** | stdio | streamable-http | `/mcp` | ✅ PASS |
| Edge case | stdio | sse | `/sse#name` | ✅ PASS |
| Default | stdio | "" (empty) | `/mcp` | ✅ PASS |
| Unchanged | sse | (ignored) | `/sse#name` | ✅ PASS |
| Unchanged | streamable-http | (ignored) | `/mcp` | ✅ PASS |

All 15 tests pass including targetURI path handling.

---

## Changes

**7 files changed:**

- `pkg/transport/url.go` - Added proxyMode parameter and logic
- `pkg/transport/url_test.go` - Added comprehensive test coverage
- `cmd/thv-operator/controllers/mcpserver_controller.go` - Updated call site
- `pkg/runner/runner.go` - Updated call site with type conversion
- `pkg/workloads/manager.go` - Updated call site with type conversion  
- `pkg/workloads/types/types.go` - Updated call site with empty string default
- `pkg/vmcp/workloads/k8s.go` - Updated call site with type conversion

**Build status:**
```bash
$ go build ./...
```

---

## Impact

This fix ensures stdio transport respects the `proxyMode` setting:

| Config | Before (Bug) | After (Fixed) |
|--------|--------------|---------------|
| stdio + streamable-http | `/sse` ❌ | `/mcp` ✅ |
| stdio + sse | `/sse` ✅ | `/sse` ✅ |
| stdio + (empty) | `/sse` ❌ | `/mcp` ✅ |
| sse transport | `/sse` ✅ | `/sse` ✅ |
| streamable-http transport | `/mcp` ✅ | `/mcp` ✅ |

Soooo since CRD is default for `proxyMode is `streamable-http` bug affects the default stdio configuration

---

Fixes #2920

### CODE REVIEW

File: pkg/workloads/types/types.go [hunk 1]
```diff
@@ -72,7 +72,7 @@ func WorkloadFromContainerInfo(container *runtime.ContainerInfo) (core.Workload,
 	// Generate URL for the MCP server
 	url := ""
 	if port > 0 {
-		url = transport.GenerateMCPServerURL(transportType, transport.LocalhostIPv4, port, name, "")
+		url = transport.GenerateMCPServerURL(transportType, "", transport.LocalhostIPv4, port, name, "")
```
Comment (jhrozek) [inline pkg/workloads/types/types.go]: This is passing an empty string for `proxyMode`, but the actual value from `runConfig.ProxyMode` isn't loaded until line 93. The URL gets generated before we know what the proxy mode is.

If you run `thv run` with `--transport stdio --proxy-mode sse` and then `thv list`, you'll see the URL shows `/mcp` instead of `/sse#name`.

The fix is to move the `loadRunConfigFields` call before the URL generation:

```go
transportType := labels.GetTransportType(container.Labels)

tType, err := types.ParseTransportType(transportType)
if err != nil {
    tType = types.TransportTypeSSE
}

// Load RunConfig first so we have proxyMode available
ctx := context.Background()
runConfig, err := loadRunConfigFields(ctx, name)
if err != nil {
    return core.Workload{}, err
}

// Now we can generate the URL with the correct proxyMode
url := ""
if port > 0 {
    url = transport.GenerateMCPServerURL(transportType, runConfig.ProxyMode, transport.LocalhostIPv4, port, name, "")
}
```

File: pkg/transport/url_test.go [hunk 2]
```diff
@@ -105,18 +135,39 @@ func TestGenerateMCPServerURL(t *testing.T) {
...
+			targetURI:     "http://remote.com/api",
+			expected:      "http://localhost:12345/api#test-container",
+		},
 	}
 
 	for _, tt := range tests {
 		t.Run(tt.name, func(t *testing.T) {
 			t.Parallel()
-			url := GenerateMCPServerURL(tt.transportType, tt.host, tt.port, tt.containerName, tt.targetURI)
+			url := GenerateMCPServerURL(tt.transportType, tt.proxyMode, tt.host, tt.port, tt.containerName, tt.targetURI) // ✅ ADD proxyMode parameter
```
Comment (jhrozek) [inline pkg/transport/url_test.go]: Leftover comment from development - please remove before merging.

```suggestion
			url := GenerateMCPServerURL(tt.transportType, tt.proxyMode, tt.host, tt.port, tt.containerName, tt.targetURI)
```

File: pkg/transport/url.go [hunk 3]
```diff
@@ -15,11 +15,33 @@ import (
...
-	isSSE := transportType == types.TransportTypeSSE.String() || transportType == types.TransportTypeStdio.String()
-	isStreamable := transportType == types.TransportTypeStreamableHTTP.String()
+	var isSSE, isStreamable bool
+
+	if transportType == types.TransportTypeStdio.String() {
+		// For stdio, the proxy mode determines the HTTP endpoint
+		// Default to streamable-http if proxyMode is empty (matches CRD default)
+		effectiveProxyMode := proxyMode
+		if effectiveProxyMode == "" {
+			effectiveProxyMode = "streamable-http"
```
Comment (jhrozek) [inline pkg/transport/url.go]: Minor: consider using the typed constant `types.ProxyModeStreamableHTTP.String()` instead of the magic string. Same for `"sse"` on line 32. Not a blocker, but it's more consistent with the rest of the codebase.

**Comment by slyt3** [comment]
Thankyou for your response, i will look into it and fix

File: pkg/workloads/types/types.go (hunk 1 above)
Comment (slyt3) [reply pkg/workloads/types/types.go]: Earlier i tried it and its still doesnt work, i will look on this issue and will fix it until tommorow morning

Comment (jhrozek) [reply pkg/workloads/types/types.go]: if you get stuck, that's alright. I'll be also fine pushing into the PR myself (you'd still retain the attribution) so you don't have to bang your had against the wall

Comment (slyt3) [reply pkg/workloads/types/types.go]: Its fine i will fix it, thats the point of learning, you get stuck and search for solutions and I love it

**Comment by slyt3** [comment]
@jhrozek Fixed the issues from code review

Tested end-to-end:
- `./thv run --transport stdio --proxy-mode sse fetch` now correctly generates `/sse#fetch` URL
- Checked lint before pushing it had 0 issues
- And all changed code passed tests.

Sorry for the initial confusion - I forgot to rebuild the binary after the first fix! 

The CI errors appear to be from stale cache. Fresh clone builds successfully: go build ./... passes with no errors. The files mentioned (file_status.go:152, k8s.go:380) don't have GenerateMCPServerURL calls at those line numbers in the current code.

**Comment by slyt3** [comment]
nvm gonna sync with the base branch and fix everything

**Comment by slyt3** [comment]
Fixed the issues that we had with base branch.  Waiting for @jhrozek your review.

**Review by jhrozek — APPROVED** [review]
I've done extensive testing on this PR and it looks good.

The bug was real - before this fix, stdio transport servers with `proxyMode: streamable-http` were getting SSE URLs (`/sse#name`) instead of the correct `/mcp` endpoint. This caused a mismatch between what clients expected and what they got.

**What I tested:**

Deployed the fix to a Kind cluster and verified:
- `stdio` + `proxyMode: streamable-http` now correctly generates `/mcp` (was `/sse#...` before)
- `stdio` + `proxyMode: sse` still generates `/sse#name` as expected
- Native SSE transport servers are unaffected

Also confirmed this isn't a breaking change - existing workloads already have `proxyMode` saved in their specs, so their URLs won't suddenly change.

All unit tests pass. The fix is straightforward and touches all the right callsites.
